### PR TITLE
Blended node pooling

### LINKED ISSUE #2551: The parser should have caps on the BlendedNode[] pool
Context:

> http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Parser/SyntaxParser.cs,32

The parser caches `BlendedNode[]` allocations using our standard object pools.  This particular array is used for operations like look ahead in the parser.  The array is dynamically grown on demand to accommodate the desired look ahead length. 

When the source file contains bad code this array can get incredibly large.  One such internal bug report sees the array at 2 ^ 24 elements.  In such scenarios we should not be returning the object to the pool.  It just serves to increase the memory footprint of the process.  I'm unsure what the cap should be but 2 ^ 24 seems to be above it ;) 

Yes we should definitely focus on fixing the look ahead issues to not allocate so many elements.  But there will always be code that tricks the parser into going down bad paths and we should do our best to recover from those scenarios.

### PR DESCRIPTION
**Pull request by pharring**
Limit the size of BlendedNode arrays returned to the pool in C# incremental parsing. The 4096 value was determined by gathering statistics on typing scenarios in a reasonably large solution. The pathological case described in #2551 has over 2^19 elements.
Fixes #2551

### CODE REVIEW

**Comment by pharring** [comment]
@jaredpar @VSadov @gafter for review.

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs [hunk 1]
```diff
@@ -76,9 +76,16 @@ public void Dispose()
             var blendedTokens = _blendedTokens;
             if (blendedTokens != null)
             {
-                Array.Clear(_blendedTokens, 0, _blendedTokens.Length);
-                s_blendedNodesPool.Free(_blendedTokens);
                 _blendedTokens = null;
+                Array.Clear(blendedTokens, 0, blendedTokens.Length);
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: no need to clear when not returning to the pool

**Comment by VSadov** [comment]
LGTM

**Comment by gafter** [comment]
LGTM

**Comment by jaredpar** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs (hunk 1 above)
Comment (pharring) [inline src/Compilers/CSharp/Portable/Parser/SyntaxParser.cs]: ok

**Comment by pharring** [comment]
@dotnet-bot retest this, please.

**Comment by pharring** [comment]
@dotnet-bot retest this, pretty please.

**Comment by pharring** [comment]
@dotnet-bot retest this.

**Comment by pharring** [comment]
@dotnet-bot test this please

**Pull request merged**